### PR TITLE
refactor!: remove template usage from grid header and footer

### PR DESCRIPTION
**Pull request by sissbruecker**
## Description

Removes the usage of `TemplateRenderer` and `ComponentRenderer` for rendering header and footer content in grid columns, and instead adds connector methods to define header/footer renderer functions on the client grid column that render the respective text or component content. This effectively removes any usage of the `<template>` element for grid headers and footers.

Breaking changes:
- Header and footer do not contain a `<flow-component-renderer>` tag anymore when using component header or footer
- removes protected `AbstractColumn.setHeaderRenderer`
- removes protected `AbstractColumn.getHeaderRenderer`
- removes protected `AbstractColumn.setFooterRenderer`
- removes protected `AbstractColumn.getFooterRenderer`

Closes https://github.com/vaadin/flow-components/issues/3574
Fixes https://github.com/vaadin/web-components/issues/3470

## Type of change

- Refactoring

### CODE REVIEW

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridViewFilteringIT.java [hunk 1]
```diff
@@ -66,17 +65,14 @@ public void scrollToEnd_filter_rowsUpdated() {
...
 
         String html = headerCell.getInnerHTML();
         if (withSorter) {
             Assert.assertTrue(html.contains("<vaadin-grid-sorter"));
         } else {
             Assert.assertFalse(html.contains("<vaadin-grid-sorter"));
         }
-        if (componentRenderer) {
-            Assert.assertTrue(html.contains("<flow-component-renderer"));
-        }
```
Comment (sissbruecker) [inline vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridViewFilteringIT.java:79]: Cells do not contain a `flow-component-renderer` tag anymore, remaining assertions should still verify that a component tag is rendered.

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridViewHeaderAndFooterRowsIT.java [hunk 2]
```diff
@@ -77,48 +68,43 @@ public void gridWithHeaderWithComponentRenderer_headerAndFooterAreRenderered() {
...
 
         String html = headerCell.getInnerHTML();
         if (withSorter) {
             Assert.assertTrue(html.contains("<vaadin-grid-sorter"));
         } else {
             Assert.assertFalse(html.contains("<vaadin-grid-sorter"));
         }
-        if (componentRenderer) {
-            Assert.assertTrue(html.contains("<flow-component-renderer"));
-        }
```
Comment (sissbruecker) [inline vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridViewHeaderAndFooterRowsIT.java:114]: Cells do not contain a `flow-component-renderer` tag anymore, remaining assertions should still verify that a component tag is rendered.

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/SortingIT.java [hunk 3]
```diff
@@ -79,9 +79,8 @@ public void setInitialSortOrder_changeOrderFromServer_sortIndicatorsUpdated() {
     public void setInitialSortOrder_sorterAriaLabels() {
         findElement(By.id("sort-by-age")).click();
         List<TestBenchElement> sorters = grid.$("vaadin-grid-sorter").all();
-        // The first sorter uses a ComponentRenderer so an aria-label can't be
-        // generated
-        Assert.assertEquals(false, sorters.get(0).hasAttribute("aria-label"));
+        Assert.assertEquals("Sort by Name",
+                sorters.get(0).getAttribute("aria-label"));
```
Comment (sissbruecker) [inline vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/SortingIT.java:83]: The new connector method for rendering the header does fix: https://github.com/vaadin/web-components/issues/3470

Comment (knoobie) [reply vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/SortingIT.java:83]: Thank you 🙇‍♂️

**Review by web-padawan — APPROVED** [review]

**Review by ugur-vaadin — APPROVED** [review]

**Pull request merged**